pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -244,86 +244,84 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
 		return opcontroller.WaitForRequeueIfCephClusterNotReady, nil
 	}
 
-	switch {
-	// osd is down
-	case osdDown:
-		logger.Infof("osd is down in failure domain %q. pg health: %q", drainingFailureDomain, pgHealthMsg)
-		currentlyDrainingFD, ok := pdbStateMap.Data[drainingFailureDomainKey]
-		if !ok || drainingFailureDomain != currentlyDrainingFD {
-			pdbStateMap.Data[drainingFailureDomainKey] = drainingFailureDomain
-			pdbStateMap.Data[drainingFailureDomainDurationKey] = time.Now().Format(time.RFC3339)
-		}
-		if activeNodeDrains {
-			pdbStateMap.Data[setNoOut] = "true"
-		}
+	osdDown := len(downOSDs) > 0
+	maxUnavailableOSDCount := 0
 
-	// osd is back up and either pgs have become healthy or pg healthy check timeout has elapsed
-	case !osdDown && (pgClean || r.hasPGHealthCheckTimedout(pdbStateMap)):
-		// reset the configMap if cluster is clean or if the timeout for PGs to become active+clean has exceeded
-		logger.Debugf("no OSD is down in the %q failure domains: %v. pg health: %q", failureDomainType, allFailureDomains, pgHealthMsg)
+	// switch block to update the PDB state config map based on the PG status and running OSDs
+	switch {
+	case !osdDown && pgClean:
+		logger.Infof("OSDs are up and PGs are clean. PG status: %q", pgHealthMsg)
+		maxUnavailableOSDCount = 1
 		resetPDBConfig(pdbStateMap)
 
-	default:
-		logger.Infof("all %q failure domains: %v. osd is down in failure domain: %q. active node drains: %t. pg health: %q", failureDomainType,
-			allFailureDomains, drainingFailureDomain, activeNodeDrains, pgHealthMsg)
-	}
-
-	if pdbStateMap.Data[setNoOut] == "true" {
-		err = r.updateNoout(clusterInfo, pdbStateMap, allFailureDomains)
-		if err != nil {
-			logger.Errorf("failed to update maintenance noout in cluster %q. %v", request, err)
+	case osdDown && pgClean:
+		logger.Infof("OSD(s) %v are down but PGs are clean. PG Status: %q", downOSDs, pgHealthMsg)
+		// In case of a node drain event, the OSD pods can get drained rapidly and it would take some time for rook to fetch
+		// the correct PG status. So wait for 60 seconds when OSD is down and node drain event is detected
+		if len(nodeDrainFailureDomains) > 0 {
+			lastNodDrainTimeStamp, err := getLastNodeDrainTimeStamp(pdbStateMap, drainingFailureDomainDurationKey)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrapf(err, "failed to get last node drain timestamp from the configmap %q", pdbStateMap.Name)
+			}
+			if time.Since(lastNodDrainTimeStamp) < 60*time.Second {
+				logger.Infof("node drain is detected. Requeue to ensure that correct PG status is read.")
+				maxUnavailableOSDCount = 1
+			} else {
+				maxUnavailableOSDCount = len(downOSDs) + 1
+				resetPDBConfig(pdbStateMap)
+			}
+		} else {
+			maxUnavailableOSDCount = len(downOSDs) + 1
```

Comment:
Where do we check that the down OSDs also have 0 PGs?

Reply:
@travisn 
We don't. I couldn't find a reliable/easy way to do that. So currently: 
- If OSDs go down, we wait for 30 seconds before checking the PG healthy. 
- If PGs are healthy even after 30 seconds, then we consider that PG didn't get affected due to the OSDs going down. So then we change the `maxunavailable` to `1+downOSDs`
Note: Rook used to have this same`wait` logic before the last upstream changes made by Ushitaro. So not doing anything new here, just using the same wait logic again.